src/storage/mvcc/reader.rs

```diff
@@ -154,24 +163,62 @@ impl<'a> MvccReader<'a> {
         Ok(None)
     }
 
-    fn create_data_cursor(&mut self) -> Result<()> {
-        if self.data_cursor.is_none() {
-            self.data_cursor = Some(try!(self.snapshot
-                .iter(None, self.scan_mode.unwrap_or(ScanMode::Mixed))));
+    fn create_write_cursor(&mut self) -> Result<()> {
+        if self.write_cursor.is_none() {
+            self.write_cursor = Some(try!(self.snapshot
+                .iter_cf(CF_WRITE, None, self.scan_mode.unwrap_or(ScanMode::Mixed))));
+        }
+        Ok(())
+    }
+
+    fn create_lock_cursor(&mut self) -> Result<()> {
+        if self.lock_cursor.is_none() {
+            self.lock_cursor = Some(try!(self.snapshot
+                .iter_cf(CF_LOCK, None, self.scan_mode.unwrap_or(ScanMode::Mixed))));
         }
         Ok(())
     }
 
     pub fn seek(&mut self, mut key: Key, ts: u64) -> Result<Option<(Key, Value)>> {
-        try!(self.create_data_cursor());
+        assert!(self.scan_mode.is_some());
+        try!(self.create_write_cursor());
+        try!(self.create_lock_cursor());
+
+        let (mut write_valid, mut lock_valid) = (true, true);
 
         loop {
             key = {
-                let mut cursor = self.data_cursor.as_mut().unwrap();
-                if !try!(cursor.near_seek(&key)) {
-                    return Ok(None);
+                let mut w_cur = self.write_cursor.as_mut().unwrap();
+                let mut l_cur = self.lock_cursor.as_mut().unwrap();
+                let (mut w_key, mut l_key) = (None, None);
+                if write_valid {
+                    if try!(w_cur.near_seek(&key)) {
+                        w_key = Some(w_cur.key());
+                    } else {
+                        w_key = None;
+                        write_valid = false;
+                    }
+                }
+                if lock_valid {
+                    if try!(l_cur.near_seek(&key)) {
+                        l_key = Some(l_cur.key());
+                    } else {
+                        l_key = None;
+                        lock_valid = false;
+                    }
+                }
+                match (w_key, l_key) {
+                    (None, None) => return Ok(None),
+                    (None, Some(k)) => Key::from_encoded(k.to_vec()),
+                    (Some(k), None) => try!(Key::from_encoded(k.to_vec()).truncate_ts()),
+                    (Some(wk), Some(lk)) => {
+                        if wk < lk {
```

Comment:
There is no easy way to reuse lk without allocate a new vector. This may add overhead to ScanMode::Forward at every seek.